rusk-wallet/src/clients.rs

```diff
@@ -248,133 +135,181 @@ impl StateStore {
-        Arc::clone(&state.cache)
-    }
-}
+            let prove_req = RuskRequest::new("prove_execute", proof.to_vec());
+
+            let proof =
+                prover.call(2, "rusk", &prove_req).wait().map_err(|e| {
+                    ExecutionCoreError::PhoenixCircuit(e.to_string())
+                })?;
+
+            tx.set_proof(proof);
+
+            status("Proving sucesss!");
+        }
 
-/// Types that are clients of the state API.
-impl StateClient for StateStore {
-    /// Error returned by the node client.
-    type Error = Error;
+        let tx_bytes = utx.to_var_bytes();
+
+        status("Attempt to preverify tx...");
+        let preverify_req = RuskRequest::new("preverify", tx_bytes.clone());
+        let _ = self.client.call(2, "rusk", &preverify_req).wait()?;
+        status("Preverify success!");
+
+        status("Propagating tx...");
+        let propagate_req = RuskRequest::new("propagate_tx", tx_bytes);
+        let _ = self.client.call(2, "Chain", &propagate_req).wait()?;
+        status("Transaction propagated!");
+
+        Ok(utx)
+    }
 
     /// Find notes for a view key, starting from the given block height.
-    fn fetch_notes(
+    pub(crate) fn inputs(
         &self,
-        vk: &ViewKey,
-    ) -> Result<Vec<EnrichedNote>, Self::Error> {
-        let psk = vk.public_spend_key();
-        let state = self.inner.lock().unwrap();
-
-        Ok(state
-            .cache
-            .notes(&psk)?
+        index: u8,
+        target: u64,
+    ) -> Result<Vec<(Note, NoteOpening, BlsScalar)>, Error> {
+        let vk = derive_phoenix_vk(self.store().get_seed(), index);
+        let sk = derive_phoenix_sk(self.store().get_seed(), index);
+        let pk = derive_phoenix_pk(self.store().get_seed(), index);
+
+        let inputs: Result<Vec<_>, Error> = self
+            .cache()
+            .notes(&pk)?
             .into_iter()
-            .map(|data| (data.note, data.height))
-            .collect())
-    }
+            .map(|data| {
+                let note = data.note;
+                let nullifiers = note.gen_nullifier(&sk);
+                let value = note.value(Some(&vk)).unwrap();
 
-    /// Fetch the current anchor of the state.
-    fn fetch_anchor(&self) -> Result<BlsScalar, Self::Error> {
-        let state = self.inner.lock().unwrap();
+                Ok((note, value, nullifiers))
+            })
+            .collect();
 
-        self.status("Fetching anchor...");
+        try_input_notes(inputs?, target)
+            .into_iter()
+            .map(|(note, scalar)| {
+                let opening = self.fetch_opening(&note)?;
 
-        let anchor = state
-            .client
-            .contract_query::<(), 0>(TRANSFER_CONTRACT, "root", &())
-            .wait()?;
-        self.status("Anchor received!");
-        let anchor = rkyv::from_bytes(&anchor).map_err(|_| Error::Rkyv)?;
-        Ok(anchor)
+                Ok((note, opening, scalar))
+            })
+            .collect()
```

Comment:
Since the secret-key is generated in this function, you need to zeroize it after using it and before it goes out of scope.

Reply:
zerorized